sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java

```diff
@@ -78,17 +78,21 @@
    */
   @Deprecated
   public static BoundedSource<Long> upTo(long numElements) {
-    checkArgument(numElements > 0, "numElements (%s) must be greater than 0", numElements);
+    checkArgument(numElements >= 0,
+        "numElements (%s) must be greater than or equal to 0",
+        numElements);
     return new BoundedCountingSource(0, numElements);
   }
 
   /**
    * Creates a {@link BoundedSource} that will produce elements
-   * from {@code startIndex} to {@code endIndex - 1}.
+   * starting from {@code startIndex} to {@code endIndex - 1}.
```

Comment:
what about:
"starting from {@code startIndex} (inclusive) to {@code endIndex} (exclusive)."